From 921df8f493b7684f4e6079d43968517c4b9e2dd3 Mon Sep 17 00:00:00 2001
From: Ethan Wei <ethanwei@ami.com>
Date: Mon, 16 Oct 2023 00:56:31 -0700
Subject: [PATCH 12/13] 
 0018-Add-VLAN_MAX_NUM-for-not-creating-VLAN-interfaces-over-size.patch

---
 src/ethernet_interface.cpp                    | 24 +++++++++++++++++++
 src/ethernet_interface.hpp                    |  6 ++
 .../Network/VLAN/Create.interface.yaml        |  1 +
 3 files changed, 31 insertions(+)

diff --git a/src/ethernet_interface.cpp b/src/ethernet_interface.cpp
index 3784b8b..378c442 100644
--- a/src/ethernet_interface.cpp
+++ b/src/ethernet_interface.cpp
@@ -58,6 +58,8 @@ constexpr auto sysctlConfigSurffix = "/arp_ignore";
 std::string arpResponseDisable = "echo 2 >";
 std::string arpResponseEnable = "echo 0 >";
 
+constexpr auto VLAN_MAX_NUM = 2;
+
 template <typename Func>
 inline decltype(std::declval<Func>()())
     ignoreError(std::string_view msg, stdplus::zstring_view intf,
@@ -698,6 +700,22 @@ ObjectPath EthernetInterface::createVLAN(uint16_t id)
                               Argument::ARGUMENT_VALUE(idStr.c_str()));
     }
 
+    if (auto size = getCreatedVLANNum(config::pathForIntfConf(
+            manager.get().getConfDir(), interfaceName()));
+        size >= VLAN_MAX_NUM)
+    {
+        log<level::ERR>(
+            fmt::format(
+                "There are already {} VLAN interfaces. so not create VLAN with vid {}",
+                VLAN_MAX_NUM, id)
+                .c_str());
+        elog<NotAllowed>(NotAllowedArgument::REASON(
+            fmt::format(
+                "There are already {} VLAN interfaces. so not create VLAN with vid {}",
+                VLAN_MAX_NUM, id)
+                .c_str()));
+    }
+
     auto objRoot = std::string_view(objPath).substr(0, objPath.rfind('/'));
     auto macStr = MacAddressIntf::macAddress();
     std::optional<stdplus::EtherAddr> mac;
@@ -1425,6 +1443,12 @@ bool EthernetInterface::sysctlConfig(const std::string& cmd)
     return ((outConfig.empty()) ? 1 : 0);
 }
 
+int EthernetInterface::getCreatedVLANNum(fs::path confFile)
+{
+    config::Parser config(confFile);
+    return (config.map.getValueStrings("Network", "VLAN")).size();
+}
+
 int EthernetInterface::setDwMacAddrByIP(std::string gateway)
 {
     int ret = 0;
diff --git a/src/ethernet_interface.hpp b/src/ethernet_interface.hpp
index 8063435..831c27f 100644
--- a/src/ethernet_interface.hpp
+++ b/src/ethernet_interface.hpp
@@ -325,6 +325,12 @@ class EthernetInterface : public Ifaces
      *  @return status of the shell command execution
      */
     bool sysctlConfig(const std::string& cmd);
+
+    /** @brief Get the number of created VLAN interface
+     *  @param[in] confFile - The path of NetIntf configuration
+     *  @return status of the shell command execution
+     */
+    int getCreatedVLANNum(std::filesystem::__cxx11::path confFile);
 };
 
 } // namespace network
diff --git a/yaml/xyz/openbmc_project/Network/VLAN/Create.interface.yaml b/yaml/xyz/openbmc_project/Network/VLAN/Create.interface.yaml
index 053c172..e4fb621 100644
--- a/yaml/xyz/openbmc_project/Network/VLAN/Create.interface.yaml
+++ b/yaml/xyz/openbmc_project/Network/VLAN/Create.interface.yaml
@@ -20,3 +20,4 @@ methods:
       errors:
           - xyz.openbmc_project.Common.Error.InvalidArgument
           - xyz.openbmc_project.Common.Error.ResourceNotFound
+          - xyz.openbmc_project.Common.Error.NotAllowed
-- 
2.43.0
